crates/storage/store.rs

```diff
@@ -358,6 +363,64 @@ impl Store {
         Ok(Some(state_trie.hash()?))
     }
 
+    /// Applies account updates to the given state tries but does not commit to db
+    /// and instead returns the data that should be committed
+    pub fn apply_account_updates_without_committing(
+        &self,
+        state_trie: &mut Trie,
+        account_updates: &[AccountUpdate],
+    ) -> Result<DataToCommitAfterAccountUpdates, StoreError> {
+        let mut storage_tries = vec![];
+        let mut accounts_code = vec![];
+
+        for update in account_updates.iter() {
+            let hashed_address = hash_address(&update.address);
+            if update.removed {
+                // Remove account from trie
+                state_trie.remove(hashed_address)?;
+            } else {
+                // Add or update AccountState in the trie
+                // Fetch current state or create a new state to be inserted
+                let mut account_state = match state_trie.get(&hashed_address)? {
+                    Some(encoded_state) => AccountState::decode(&encoded_state)?,
+                    None => AccountState::default(),
+                };
+                if let Some(info) = &update.info {
+                    account_state.nonce = info.nonce;
+                    account_state.balance = info.balance;
+                    account_state.code_hash = info.code_hash;
+                    // Store updated code in DB
+                    if let Some(code) = &update.code {
+                        accounts_code.push((info.code_hash, code.clone()));
+                    }
+                }
+                // Store the added storage in the account's storage trie and compute its new root
+                if !update.added_storage.is_empty() {
+                    let mut storage_trie = self.engine.open_storage_trie(
+                        H256::from_slice(&hashed_address),
+                        account_state.storage_root,
+                    );
+                    for (storage_key, storage_value) in &update.added_storage {
+                        let hashed_key = hash_key(storage_key);
+                        if storage_value.is_zero() {
+                            storage_trie.remove(hashed_key)?;
+                        } else {
+                            storage_trie.insert(hashed_key, storage_value.encode_to_vec())?;
+                        }
+                    }
+                    account_state.storage_root = storage_trie.hash_no_commit();
+                    storage_tries.push((H256::from_slice(&hashed_address), storage_trie));
+                }
+                state_trie.insert(hashed_address, account_state.encode_to_vec())?;
+            }
+        }
+
+        Ok(DataToCommitAfterAccountUpdates {
+            storage_tries,
+            accounts_code,
+        })
```

Comment:
Idea: see if we can sort these, it might help locality when inserting to the DB.